homeassistant/components/obihai/config_flow.py

```diff
@@ -43,20 +46,59 @@ class ObihaiFlowHandler(ConfigFlow, domain=DOMAIN):
 
     VERSION = 1
 
+    def __init__(self) -> None:
+        """Initialize."""
+        self._host: str | None = None
+
     async def async_step_user(
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
         """Handle a flow initialized by the user."""
         errors: dict[str, str] = {}
 
         if user_input is not None:
-            self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
-            if await async_validate_creds(self.hass, user_input):
-                return self.async_create_entry(
-                    title=user_input[CONF_HOST],
-                    data=user_input,
-                )
-            errors["base"] = "cannot_connect"
+            try:
+                ip = gethostbyname(user_input[CONF_HOST])
+                await self.async_set_unique_id(get_mac_address(ip=ip))
+                self._abort_if_unique_id_configured()
+
+                if await async_validate_creds(self.hass, user_input):
+                    return self.async_create_entry(
+                        title=user_input[CONF_HOST],
+                        data=user_input,
+                    )
+                errors["base"] = "invalid_auth"
+            except gaierror:
+                errors["base"] = "cannot_connect"
+
+        user_input = {CONF_HOST: self._host if self._host else ""}
+        data_schema = self.add_suggested_values_to_schema(DATA_SCHEMA, user_input)
+        return self.async_show_form(
+            step_id="user",
+            errors=errors,
+            data_schema=data_schema,
+        )
+
+    async def async_step_dhcp(self, discovery_info: dhcp.DhcpServiceInfo) -> FlowResult:
+        """Prepare configuration for a DHCP discovered Obihai."""
+        errors: dict[str, str] = {}
+        self._host = discovery_info.ip
+
+        await self.async_set_unique_id(discovery_info.macaddress)
+        self._abort_if_unique_id_configured()
```

Comment:
Existing entries for existing users don't have a unique ID set. So, their existing instances will now be discovered again?

Reply:
Good point.  Uploaded a commit that migrates unique_ids to mac address, though it sounds like we still need to close on what an appropriate unique ID is.  Serial number requires authentication, and MAC may not be ideal in certain network topologies.

I just uploaded two commits with two different ways of handling this.  The first just uses get_mac_address to migrate unique IDs.  The second uses the Obihai API to get MAC - My concern here is that if get_mac_address is returning an incorrect value these values won't match, but at least async_setup_entry will attempt to migrate the unique ID.  I'm not sure if/how this would fail in practice and don't really have an environment I could test this thoroughly in...